src/openms/source/CHEMISTRY/Tagger.cpp

```diff
@@ -19,36 +19,42 @@
 
 namespace OpenMS
 {
-  char Tagger::getAAByMass_(double m) const
+  // Define a method to get amino acid by mass
+char Tagger::getAAByMass_(double m) const
   {
     // fast check for border cases
-    if (m < min_gap_ || m > max_gap_) return ' ';
-
+    if (m < min_gap_ || m > max_gap_) return ' '; // If mass is out of the valid range, return a space character
+    // Calculate delta using a function from the Math class, converting ppm to mass
     const double delta = Math::ppmToMass(ppm_, m);
+    // Find the lower bound of m - delta in mass2aa
     auto left = mass2aa_.lower_bound(m - delta);
     //if (left == mass2aa_.end()) return ' '; // cannot happen, since we checked boundaries above
-
-    if (fabs(left->first - m) < delta) return left->second;
+  // Check if the absolute difference between left->first and m is smaller than delta
+    if (fabs(left->first - m) < delta) return left->second; // If yes, return the associated amino acid
```

Comment:
if ```m=3.001``` then ```auto left = mass2aa_.lower_bound(m - delta);```
will give you 3 because we subtract delta.
So I think it will give you the first hit. Because AA masses are quite different, this hit should be unique.
Or did I get it wrong?

Reply:
right, it depends on the delta, but you may not get the optimal result, i.e. closest to the query mass
The argument still holds :)

Reply:
the delta is the tolerance allowed for a valid hit here.
so in your example, if `lower_bound(m - delta)` returns 4, then `m - delta` > 3 and therefore 3 is outside of the tolerance window. 4 will also be, and that is caught by `fabs(left->first - m) < delta`
As Timo mentioned, the AA masses are different enough that there should only be one AA candidate. Either it is within the tolerance (`(3.001 - delta) < 3.0` and `(3 - 3.001) < delta`) or it is not (`(3.001 - delta) < 4.0` but `(4 - 3.001) > delta`).
Returning 3 for m = 3.001 only makes sense, if the tolerance `delta >= 0.001`. So either 3 is returned, or nothing (space string), depending on delta.
The goal is not getting the AA that is the closest to the query mass, but an AA that is close enough to fall within the tolerance window, or otherwise nothing to end the current sequence.